[FLINK-8639][State Backends]Fix always need to seek multiple times when iterator RocksDBMapState

sihuazhou opened the pull request:
## What is the purpose of the change

This PR addressed issue [FLINK-8639](https://issues.apache.org/jira/browse/FLINK-8639). Currently, almost every time we want to iterator a RocksDBMapState we need to do seek at least 2 times (Seek is a poor performance action for rocksdb cause it can't use the bloomfilter). This is because `RocksDBMapIterator` use a `cacheEntries` to cache the seek values every time and the `cacheEntries`'s init size is 1.

## Brief change log


- Change CACHE_SIZE_BASE from 1 to 32.
- Compare in reverse order in function underSameKey().

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

sihuazhou commented:
Hi, @StefanRRichter cloud you please have a look at this?

sihuazhou commented:
I think maybe a best way to fix this is to make `mapState.iterator()` return a `FlinkIterator` , aand `FlinkIterator` implement `AutoCloseable`, and let `FlinkIterator` hold the `RocsdbIterator` until it closed. This way we only need to seek one time when we iterate a map state. But this will broke the interface flink has provided.  @StefanRRichter What do you think?

StefanRRichter commented on flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java:
```diff
@@ -319,6 +321,9 @@ private UV deserializeUserValue(byte[] rawValueBytes) throws IOException {
 		private byte[] rawValueBytes;
 
+		/** The offset of User Key offset in raw key bytes. */
+		private final int userKeyOffset;
```
This value is basically a constant on a per-backend level and we could remember the value in the MapState itself, no need to replicate it into every entry. It can then still be recalled from there for `deserializeUserKey(...)`.

sihuazhou replied:
👍 addressed.

StefanRRichter commented on flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java:
```diff
@@ -520,7 +530,7 @@ private boolean underSameKey(byte[] rawKeyBytes) {
 			}
 
-			for (int i = 0; i < keyPrefixBytes.length; ++i) {
+			for (int i = keyPrefixBytes.length - 1; i >= backend.getKeyGroupPrefixBytes(); --i) {
```
nit: maybe `for (int i = keyPrefixBytes.length; --i >= backend.getKeyGroupPrefixBytes();)` looks a bit cleaner.

sihuazhou replied:
👍 addressed.

StefanRRichter commented on flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java:
```diff
@@ -400,7 +410,7 @@ public UV setValue(UV value) {
 	private abstract class RocksDBMapIterator<T> implements Iterator<T> {
 
-		static final int CACHE_SIZE_BASE = 1;
+		static final int CACHE_SIZE_BASE = 32;
```
Why is 32 the magic value and not 16 or 64? Just asking :)

StefanRRichter replied:
And if memory is really a concern, then I would also think about releasing the raw bytes from the entry once we have lazily deserialized the objects from them.

sihuazhou replied:
Indeed, I don't think the memory is really a concern (at least when CACHE_SIZE_LIMIT=128), I almost like to change the CACHE_SIZE to a fixed value (like 128), What do you think?

StefanRRichter commented:
About your `AutoCloseable` proposal, I think on the one hand it is nicer for the implementation and on the other hand it is a bit more dangerous for users. If user code forgets to close the iterator, they can potentially not release native resources .

sihuazhou commented:
Thanks for your comment for the `AutoCloseable` proposal, I agreed that the disadvantages are more than advantages....

sihuazhou commented:
I have addressed all the comments. One thing that I am not 100 sure: I change the size of the `cacheEntries` to the fixed value `CACHE_SIZE_LIMIT` (128). @StefanRRichter Could you please have a look again?

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java:
```diff
@@ -520,7 +527,7 @@ private boolean underSameKey(byte[] rawKeyBytes) {
 			}
 
-			for (int i = keyPrefixBytes.length - 1; i >= backend.getKeyGroupPrefixBytes(); --i) {
+			for (int i = keyPrefixBytes.length - 1; --i >= backend.getKeyGroupPrefixBytes(); ) {
```
If you want to change like I proposed, I think you need to also remove the `- 1` from the initialization of i. This looks like a bug now.

sihuazhou replied:
You are right, I have addressed this like your proposed.

StefanRRichter commented:
While `128` is maybe already up to 128x better, a scientific approach for finding the constant would be to try out for very small entries from which point you have diminishing returns for higher values. Even for pretty large entries 128 should still be small in total terms. Maybe we can go higher, or maybe that is already diminishing returns. If you want, you could measure it. If you don't have the time, then 128 maybe ok for now.

sihuazhou commented:
Aha, for now I would like to leave it to 128 ...  Even I can't find a best value, but this value is not a bad value and is also a safety value.. If I found any another value that is better than it (on production), I will feed back to community at that time.

StefanRRichter commented:
Thanks for this nice contribution, changes look good now. Will merge this.
